### DATA ANALYST TEST
Author: Ashkan Moradi \
Date: November 2025

Hi. let's Import our libraries:

In [36]:
import pandas as pd

Now let's load out csv files into dataframes:

In [37]:
orders = pd.read_csv("orders.csv")
order_items = pd.read_csv("order_items.csv")
products = pd.read_csv("products.csv")

Now let's do some Exploratory Data Analysis (EDA) and take a brief look into our data. \
We take a look into five first rows, column types, and Number of rows and columns in each table. \
Let's start with order table:

In [38]:
# orders table:
print("First five rows:") 
print(orders.head(5))

print("\n Data Types:")
print(orders.dtypes)

print("\n Number of rows and columns:") 
print(orders.shape)

First five rows:
   order_id  customer_id  order_date  total_amount
0     34264       714477  2025-09-01        1295.0
1     34266       553176  2025-09-01        2685.0
2     34268       711383  2025-09-01         895.0
3     34270       714497  2025-09-01        1295.0
4     34271       714451  2025-09-01         895.0

 Data Types:
order_id          int64
customer_id       int64
order_date       object
total_amount    float64
dtype: object

 Number of rows and columns:
(7432, 4)


Now Let's take a look at order_items table:

In [39]:
# order_items table:
print("First five rows:") 
print(order_items.head(5))

print("\n Data Types:")
print(order_items.dtypes)

print("\n Number of rows and columns:") 
print(order_items.shape)

First five rows:
   order_id  product_id  quantity  unit_price   sales  discounts
0     34264          18         1       550.0   550.0      550.0
1     34264           5         1      1295.0  1295.0        0.0
2     34266          18         3       550.0  1650.0     1650.0
3     34266           3         3       895.0  2685.0        0.0
4     34268          18         1       550.0   550.0      550.0

 Data Types:
order_id        int64
product_id      int64
quantity        int64
unit_price    float64
sales         float64
discounts     float64
dtype: object

 Number of rows and columns:
(15042, 6)


And now Let's take a look at products table:

In [40]:
# products table:
print("First five rows:") 
print(products.head(5))

print("\n Data Types:")
print(products.dtypes)

print("\n Number of rows and columns:") 
print(products.shape)

First five rows:
   product_id  category          product_name unit_cost
0          -1   Unknown               Unknown         0
1           1  Mattress              Mattress   $232.20
2           2  Mattress        Plush Mattress   $309.60
3           3  Mattress         Firm Mattress   $227.04
4           4  Mattress  Hybrid Firm Mattress   $330.24

 Data Types:
product_id       int64
category        object
product_name    object
unit_cost       object
dtype: object

 Number of rows and columns:
(20, 4)


### Now We start with question #3: Data Quality Checks (part a) to check data first.

Here we check Duplicate values, Missing values (Null), and Referential integrity for each table. Let's start with order table:  

In [41]:
# 3. Data Quality Checks
# a. Identify data integrity issues

# orders table:

duplicate_orderIdAndCustomerId = orders.duplicated(subset=['order_id','customer_id'],keep=False)
duplicate_orderIdAndCustomerId_rows = orders[duplicate_orderIdAndCustomerId]
print("\nOrders with duplicate order_id and customer_id values:\n", duplicate_orderIdAndCustomerId_rows)

null_order = orders.isnull().sum()
print("\nOrdees with null value:\n", null_order)

# Check for orders referencing non-existent order_items
missing_orders = orders[~orders['order_id'].isin(order_items['order_id'])]
print("\nOrders with missing order items:\n", missing_orders)



Orders with duplicate order_id and customer_id values:
 Empty DataFrame
Columns: [order_id, customer_id, order_date, total_amount]
Index: []

Ordees with null value:
 order_id        0
customer_id     0
order_date      0
total_amount    0
dtype: int64

Orders with missing order items:
 Empty DataFrame
Columns: [order_id, customer_id, order_date, total_amount]
Index: []


Now we realized that order table does not have any duplicate row (based on order_id and customer_id values), does not have any Missing value (Null),  \
and all orders in order tables have assoociate order item in order_items table.

Note: If we had the Cutomer table, we could check the Referential integrity for cutomer_id column too, but we do not have it.  

Now Let's check Duplicate values, Missing values (Null), and Referential integrity in order_items table:

In [42]:
# 3. Data Quality Checks
# a. Identify data integrity issues

# order_items table:

duplicate_productIdAndOrderId = order_items.duplicated(subset=['order_id', 'product_id'],keep=False)
duplicate_productIdAndOrderId_rows = order_items[duplicate_productIdAndOrderId]
print("\nOrder items with duplicate order_id and product_id values:\n", duplicate_productIdAndOrderId_rows)

null_orderItem = order_items.isnull().sum()
print("\nOrder items with null value:\n", null_orderItem)

# Check for order_items referencing non-existent orders
missing_orders = order_items[~order_items['order_id'].isin(orders['order_id'])]
print("\nOrder items with missing orders:\n", missing_orders)

# Check for order_items referencing non-existent products
missing_products = order_items[~order_items['product_id'].isin(products['product_id'])]
print("\nOrder items with missing products:\n", missing_products)


Order items with duplicate order_id and product_id values:
        order_id  product_id  quantity  unit_price  sales  discounts
61        34305          18         1       550.0  550.0      550.0
62        34305          18         1       550.0  550.0      550.0
63        34305           1         1       895.0  895.0        0.0
64        34305           1         1       895.0  895.0        0.0
198       34383          16         1       155.0  155.0       46.5
...         ...         ...       ...         ...    ...        ...
14956     43963           1         1       895.0  895.0        0.0
15008     44003          18         1       550.0  550.0      550.0
15009     44003          18         1       550.0  550.0      550.0
15035     44021          18         1       550.0  550.0      550.0
15036     44021          18         1       550.0  550.0      550.0

[686 rows x 6 columns]

Order items with null value:
 order_id      0
product_id    0
quantity      0
unit_price    0
sale

Now we realized that order_items table has 686 duplicate rows (based on order_id and product_id values), \
 but it does not have any Missing value (Null), and all orders items in order_items tables have assoociate order and product in order table and product table.

Now Let's check Duplicate values, Missing values (Null), and Referential integrity in products table:

In [43]:
# 3. Data Quality Checks
# a. Identify data integrity issues

# products table:

duplicate_productId = products.duplicated(subset=['product_id'],keep=False)
duplicate_productId_rows = products[duplicate_productId]
print("\nProducts with duplicate product_id value:\n", duplicate_productId_rows)

null_Prodocts = products.isnull().sum()
print("\nProducts with null value:\n", null_Prodocts)

# Check for products referencing non-existent order_items
missing_products = products[~products['product_id'].isin(order_items['product_id'])]
print("\nProducts with missing order items:\n", missing_products)



Products with duplicate product_id value:
 Empty DataFrame
Columns: [product_id, category, product_name, unit_cost]
Index: []

Products with null value:
 product_id      0
category        0
product_name    0
unit_cost       0
dtype: int64

Products with missing order items:
    product_id category product_name unit_cost
0          -1  Unknown      Unknown         0


Now we realized that product table does not have any duplicate row (based on product_id values), Does not have any Missing value (Null), \
but one product_id (Unkonwn) in products table does not have assoociate order item in order_items table.

### Now let's start question 3: Data Quality Checks (part b) to clean data and solve these issues before doing any calculation.

Solving issue #1 (Duplicated Values in order_items table) by removing those records:

In [44]:
# 3. Data Quality Checks

# b. Handling/fixing in production:

# Removing Order items duplicate rows which has duplicate order_id and product_id values
order_items = order_items.drop_duplicates(subset=['order_id', 'product_id'],keep='first')


Solving issue #2 (Unknown product in product table) by removing that record:

In [45]:
# 3. Data Quality Checks

# b. Handling/fixing in production:

# Remove Unknown prodoct from products table:
products.drop(products[products['product_id'] == -1].index, inplace=True)
# let's check product table again:
print(products.head(5))


   product_id  category              product_name unit_cost
1           1  Mattress                  Mattress   $232.20
2           2  Mattress            Plush Mattress   $309.60
3           3  Mattress             Firm Mattress   $227.04
4           4  Mattress      Hybrid Firm Mattress   $330.24
5           5  Mattress  Hybrid Med Firm Mattress   $361.20


Solving issue #3 (Data Type mismatch in "unit_cost" column of products table) by removing prefix and change data type:

In [46]:
# Issue 3 (Data Type mismatch):
# Removing $ sign in unit cost column and changing the object (text) type into decimal (float64).
products['unit_cost'] = products['unit_cost'].str.removeprefix('$')
products['unit_cost'] = products['unit_cost'].astype('float64')

Solving issue #4 (Wrong value in "sales" column) by correcting it:

At first it seems that column 'sales' is the multiplication of 'quantity' and 'unit_price' but in some records it is not. Let's check it and correct it:

In [50]:
# Checking the 'sales' column value:
order_items['check_sales'] = order_items['quantity'] * order_items['unit_price'] == order_items['sales']
# Now let's pring the first five mismatches:
order_items[order_items['check_sales'] != True].head(5)


order_id  product_id  quantity  unit_price   sales  discounts  \
148     34351           4         1      1195.0  1295.0        0.0   
174     34368           4         1      1195.0  1295.0        0.0   
193     34378           4         1      1195.0  1295.0        0.0   
274     34420           4         1      1195.0  1295.0        0.0   
387     34485           4         1      1195.0  1295.0        0.0   

     check_sales  
148        False  
174        False  
193        False  
274        False  
387        False

In [51]:
# Correcting 'sales' column:
order_items['sales'] = order_items['quantity'] * order_items['unit_price']

### Now the data is almost clean. Let's get back to the quesion #1 part a: 

Metrics:\
gross_revenue = quantity * unit_price \
net_revenue = gross_revenue - discount \
cost = unit_cost * quantity

In [52]:
# 1. Sales Metrics

# Merge order_items with products to get category and unit_cost
order_items_merged = order_items.merge(products, on="product_id", how="left")

# a. Total revenue and total cost per product category
order_items_merged['gross_revenue'] = order_items_merged['quantity'] * order_items_merged['unit_price']
order_items_merged['net_revenue'] = (order_items_merged['quantity'] * order_items_merged['unit_price']) - order_items_merged['discounts']
order_items_merged['cost'] = order_items_merged['unit_cost'] * order_items_merged['quantity']

category_metrics = order_items_merged.groupby('category').agg(
    total_gross_revenue=('gross_revenue', 'sum'),
    total_net_revenue=('net_revenue', 'sum'),
    total_cost=('cost', 'sum')
).reset_index()

category_metrics



category  total_gross_revenue  total_net_revenue  total_cost
0  Accessory            2968692.1          583202.38   512771.18
1  Furniture             375105.0          324617.60   130021.68
2   Mattress            6517360.0         6242263.62  1720591.68

### Now Let's get back to the quesion #1 part b: 

Metrics:\
gross_margin_per =(total_gross_revenue -  total_cost)/total_gross_revenue \
net_revenue = (total_net_revenue - total_cost)/total_net_revenue 

In [53]:
# 1. Sales Metrics

# b. Gross margin % per category
category_metrics['gross_margin_perc'] = (category_metrics['total_gross_revenue'] - category_metrics['total_cost']) / category_metrics['total_gross_revenue'] * 100
category_metrics['net_margin_perc'] = (category_metrics['total_net_revenue'] - category_metrics['total_cost']) / category_metrics['total_net_revenue'] * 100

category_metrics

category  total_gross_revenue  total_net_revenue  total_cost  \
0  Accessory            2968692.1          583202.38   512771.18   
1  Furniture             375105.0          324617.60   130021.68   
2   Mattress            6517360.0         6242263.62  1720591.68   

   gross_margin_perc  net_margin_perc  
0          82.727371        12.076631  
1          65.337258        59.946201  
2          73.599867        72.436414

### Now Let's start quesion #2 part a: 

Here we need an aggregation then we shoud sort it by total_spent (descending) 

In [54]:
# 2. Customer Metrics

# a. Top 2 customers by total spending
customer_spending = orders.groupby('customer_id').agg(total_spent=('total_amount', 'sum')).reset_index()

top_customers = customer_spending.sort_values(by='total_spent', ascending=False).head(2)

top_customers

customer_id  total_spent
3680       715977      26910.0
4295       716677      20228.0

### And now question #2 part b: 

Here we should consider that each customer might have more than one order, so here we need an aggregation to calculate the average order value for each cusotmer:

In [55]:
# 2. Customer Metrics

# b. Average order value per customer
customer_avg_order = orders.groupby('customer_id').agg(avg_order_value=('total_amount', 'mean')).reset_index()

customer_avg_order

customer_id  avg_order_value
0             204           116.00
1             809          1495.00
2            2682          1235.02
3            3354           895.00
4            3403           895.00
...           ...              ...
7083       720016          2590.00
7084       720017          1635.02
7085       720018          1095.00
7086       720019          1235.02
7087       720021           795.00

[7088 rows x 2 columns]

We have done question #3 already. \
### Now Let's start quesion #4:

Here we need to import two subclasses: 

Counter => subclass for counting items. Elements are stored as dictionary keys and their counts are stored as dictionary values. \
Combination => subclass which Return successive r-length combinations of elements in the iterable.

In [56]:
from collections import Counter 
from itertools import combinations 

In [57]:
# 4. Frequently Purchased Products

# Find top 3 pairs of products purchased together
# First, group items (as a list) by order_id

order_product_groups = order_items.groupby('order_id')['product_id'].apply(list)

# Count all product pairs
pair_counter = Counter() 
for products_list in order_product_groups:
    pairs = combinations(sorted(products_list), 2)
    pair_counter.update(pairs)

# "most_common" method will list the n most common elements and their counts from the most common to the least.
top_3_pairs = pair_counter.most_common(3)
top_3_pairs

[((1, 18), 2199), ((13, 16), 980), ((16, 19), 963)]